Enforce the line break position of the ternary else

paulmelnikow opened the pull request:
I don't have a preference about whether the linebreak is before or after the `:`, but think we should pick one.

_After_ seems more consistent with, well, everything else, but I think @maxdumas had a preference to go the other way, so I'm down to hear his argument.

chenliu0831 commented:
+1 on this

maxdumas commented:
```js
condition
 ? trueBranch()
 : falseBranch()
```

That just reads better to me, more clearly highlighting which branch is which. It feels more analogous to an if-else statement. For something closer to a doomsday readability scenario:

```js
(a, b, c) => a + b - c <= 12 ?
 { a: 12, b: 'hello', c: new Thing() } :
 { a: a + b, b: new OtherThing() };
```
vs
```js
(a, b, c) => a + b - c <= 12
 ? { a: 12, b: 'hello', c: new Thing() }
 : { a: a + b, b: new OtherThing() };
```

I think in the first case it's really easy to lose the conditional operators there, especially if the individual lines are longer. If each branch takes multiple lines, I also think that it becomes a lot tougher to tell where one branch ends and the other begins:

```js
(a, b, c) => a + b - c <= 12 ?
  {
    a: 12, 
    b: 'hello', 
    c: {
      doomsday: {
        nesting: function () {
          console.log('help');
    },
  } :
  { 
    a: a + b, 
    b: new OtherThing() 
  };
```
vs
```js
(a, b, c) => a + b - c <= 12
  ? {
    a: 12, 
    b: 'hello', 
    c: {
      doomsday: {
        nesting: function () {
          console.log('help');
    },
  }
  : { 
    a: a + b, 
    b: new OtherThing() 
  };
```

The latter here very clearly reads to me like "if so, then this object, else this object".

That's my strong personal preference. That said, I'm also totally in favor of total operator consistency, so I'm ok with not making an exception here if no one else agrees.

paulmelnikow commented:
Thanks for that explanation, Max.

Here's another example which is slightly less doomsday. I include it since it might illustrate better why we'd ever want to use such a thing as a multiline ternary (rather than a plain `if`…`else`).

```js
// Option #1 (as proposed in PR)
// operator-linebreak: ["error", "after"]
const checkedElement = this.props.updated ?
    <div className="icon-container">
        <ImgAbs src="assets/check.svg" alt="checkmark" className="checkmark" />
    </div> :
    null;
```

```js
// Option #2 (Max's suggestion)
// operator-linebreak: ["error", "after", { "overrides": { "?": "before", ":": "before" }}]
const checkedElement = this.props.updated
    ? <div className="icon-container">
        <ImgAbs src="assets/check.svg" alt="checkmark" className="checkmark" />
    </div>
    : null;
```

```js
// Option #3
// operator-linebreak: ["error", "after", { "overrides": { ":": "before" }}]
const checkedElement = this.props.updated ?
    <div className="icon-container">
        <ImgAbs src="assets/check.svg" alt="checkmark" className="checkmark" />
    </div>
    : null;
```

I agree with you that `:` after the newline reads better.

But, I think I like 3 the most, because the `?` immediately follows the conditional expression. It reminds me of Ruby's boolean method names like [`File.executable?`](http://ruby-doc.org/core-1.9.3/File.html#method-c-executable-3F). In 2, it feels a bit odd to have the condition separated from the conditional operator.

dlsmith commented:
I also tend to prefer Option 2. I just find it much easier to visually parse. In practice, I don't think we tend to read these things linearly. We jump around and interpret the block as a unit. The obvious (`?`, `:`) pair is a signal that enables us to quickly recognize the block as a ternary expression and parse it accordingly.

Anyway, that's may handwavy justification for my preference, but I don't feel that strongly.

maxdumas commented:
👍
